src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java

```diff
@@ -0,0 +1,283 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.worker;
+
+import com.google.devtools.build.lib.shell.Subprocess;
+import com.google.devtools.build.lib.shell.SubprocessBuilder;
+import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse;
+import com.google.devtools.build.lib.vfs.Path;
+import java.io.File;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+
+/**
+ * An intermediate worker that receives response from the worker processes
+ */
+public class WorkerMultiplexer extends Thread {
+  private static Semaphore semInstanceMap = new Semaphore(1);
+  private static Map<Integer, WorkerMultiplexer> instanceMap = new HashMap<>();
```

Comment:
Why are these two fields constructed here and the rest inside the constructor?

What do all these fields mean? All these semaphores make me nervous. Docstrings, please.

Reply:
Docstrings added. The reason is these are used in a static function, we can't put it in a constructor.

Reply:
I moved the static variables and methods to other class `WorkerMultiplexerManager`, hopefully it makes more sense to you.